// schedule-drone/lib/util.js
// Generated by CoffeeScript 1.6.2
(function() {
  var PersistentSchedulerError, _,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    __slice = [].slice;

  _ = require('underscore');

  PersistentSchedulerError = (function(_super) {
    __extends(PersistentSchedulerError, _super);

    function PersistentSchedulerError() {
      var context, error;

      error = arguments[0], context = 2 <= arguments.length ? __slice.call(arguments, 1) : [];
      PersistentSchedulerError.__super__.constructor.call(this);
      this.name = 'PersistentSchedulerError';
      this.context = context;
      if (_.isString(error)) {
        this.message = error;
      }
      if (error instanceof Error) {
        this.message = error.message;
        this.stack = error.stack;
      }
    }

    PersistentSchedulerError.prototype.toString = function() {
      return "" + this.name + ": " + this.message + ", " + this.context + ", " + this.stack;
    };

    return PersistentSchedulerError;

  })(Error);

  exports.Error = PersistentSchedulerError;

}).call(this);
